DM-46158: Add workflow to perform schema comparisons and check for changes to deployed schemas

JeremyMcCormick opened the pull request:
This PR adds a workflow that will check for changes in schemas using the `felis diff` command. One comparison is performed using the `deepdiff` comparator, which will show all changes, including ones to metadata fields, e.g., `ivoa_ucd`. Another comparison step will use `alembic`, which will only show updates that affect the relational data model.

Finally, a comparison using the `alembic` mode will be performed on each schema from a list of "deployed" schema YAML files (which is also added by this PR). An error will be raised if any of them have changed (`dp02_dc2`, etc.).

A preliminary step in the workflow uses `git diff` to detect which files were changed. Files which were not changed in the repository need not be checked.

----

~~Blocked by https://github.com/lsst/felis/pull/109~~

This is merged now. ^^^^

gpdf commented on .github/workflows/compare.yaml:
```diff
@@ -0,0 +1,92 @@
+      - name: Get list of changed YAML files
+        working-directory: current-ref
+        run: |
+          CHANGED_FILES=$(git diff --name-only origin/main..HEAD -- ${{ env.SCHEMA_DIR }} | grep '\.yaml$')
```
This logic needs to be changed.  To avoid future surprises, it must compare the "current ref" against the base of the branch in which it appears.  This only really makes sense for PRs, because what we're trying to do is to ask the question "if this PR is merged, will it produce a schema change?".  It's true that 90%+ of the time it's the same as comparing to the tip of `main`, but not always, and we need to get this right up front.

gpdf commented on .github/workflows/compare.yaml:
```diff
@@ -0,0 +1,92 @@
+name: Compare Schemas for Changes
+
+on:
+  push:
```
This action probably only makes sense on pull requests, because what really matters is to make sure that a proposed change does not (or does, if it's expected) produce a schema change.

gpdf commented on .github/workflows/compare.yaml:
```diff
@@ -0,0 +1,92 @@
+          CHANGED_FILES=$(echo "$CHANGED_FILES" | sed "s|^${{ env.SCHEMA_DIR }}/||")
+          echo "CHANGED_FILES=$CHANGED_FILES" >> $GITHUB_ENV
+
+      - name: Check out main branch
```
This checkout needs to be changed to ensure that it's checking out the base ref of the reference branch for this PR.  That ref might not be the tip of *any* branch.  So please test that this does the right thing in such a situation.

gpdf commented on python/lsst/sdm_schemas/schemas/dp02_dc2.yaml:
Please split this file so that there is one file with only a "harmless" change (non-DB-schema-changing) and one with a schema-changing change.  This will allow verifying that the `felis diff -E -c alembic` is triggering only on the latter one.

gpdf replied:
Also please include a schema-changing change to a file that's not in the deployed list.

gpdf reviewed (changes requested):
Two significant changes:

- [x] Trigger only on PRs, and do the diff against the base ref of the PR
- [x] Modify the .yaml changes on the PR to have one file with a non-schema-changing change to a deployed schema, one with a schema-changing change to a deployed schema, and one with a schema-changing change to a non-deployed schema

Following the next review of this PR:

- [x] all three .yaml file changes must be removed from the PR before merging.

gpdf reviewed (approved):
Looks good!

Approved assuming a) "DO NOT MERGE" commits (used only for testing) are dropped and b) the commit with the `DEPLOYED.txt` file is kept separate from the workflow changes.

JeremyMcCormick commented:
I've corrected a problem where there was an error if no changes had occurred. And I have updated the workflow so that it skips subsequent skips when there are no changes. I tested these changes for both cases by also re-committing and testing on the dummy schema changes.
